### PR TITLE
Resolve absolute paths as extern under a feature flag

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/issues/44660
r? @nikomatsakis

### CODE REVIEW

File: src/librustc_resolve/resolve_imports.rs [hunk 1]
```diff
@@ -601,8 +601,60 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
     // If appropriate, returns an error to report.
     fn finalize_import(&mut self, directive: &'b ImportDirective<'b>) -> Option<(Span, String)> {
         self.current_module = directive.parent;
-
         let ImportDirective { ref module_path, span, .. } = *directive;
+
+        // Extern crate mode for absolute paths needs some
+        // special support for single-segment imports.
```
Comment (petrochenkov) [inline src/librustc_resolve/resolve_imports.rs]: The part about single-segment imports (`use my_crate;`/`use crate;`) works, but it is kinda hacked into.
Solving it properly requires larger refactoring that would also solve a number of pre-existing problems (https://github.com/rust-lang/rust/issues/29036 and similar).

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -2959,6 +2960,16 @@ impl<'a> Resolver<'a> {
                     // `$crate::a::b`
                     module = Some(self.resolve_crate_root(ident.node.ctxt));
                     continue
+                } else if i == 1 && self.session.features.borrow().extern_absolute_paths &&
+                                    path[0].node.name == keywords::CrateRoot.name() &&
```
Comment (nikomatsakis) [inline src/librustc_resolve/lib.rs]: Nit: isn't there some helper function we can use here for this test against `keywords::CrateRoot.name()`?

File: src/librustc_resolve/resolve_imports.rs [hunk 3]
```diff
@@ -601,8 +601,60 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
...
+        let extern_absolute_paths = self.session.features.borrow().extern_absolute_paths;
+        if module_path.len() == 1 && module_path[0].node.name == keywords::CrateRoot.name() {
```
Comment (nikomatsakis) [inline src/librustc_resolve/resolve_imports.rs]: Nit: as above, helper function?

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
The new gate test failed on `i686-unknown-linux-musl`.

<details>

```
`i686-unknown-linux-musl`

[00:39:46] ---- [ui] ui/feature-gate-extern_absolute_paths.rs stdout ----
[00:39:46] 	
[00:39:46] error: ui test compiled successfully!
[00:39:46] status: exit code: 0
[00:39:46] thread 'main' panicked at 'Some tests failed', /checkout/src/tools/compiletest/src/main.rs:475:21
[00:39:46] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/feature-gate-extern_absolute_paths.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--target=i686-unknown-linux-musl" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gate-extern_absolute_paths.stage2-i686-unknown-linux-musl" "-Crpath" "-O" "-Lnative=/checkout/obj/build/i686-unknown-linux-musl/native/rust-test-helpers" "-Clinker=/musl-i686/bin/musl-gcc" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/feature-gate-extern_absolute_paths.stage2-i686-unknown-linux-musl.aux" "-A" "unused"
[00:39:46] stdout:
[00:39:46] ------------------------------------------
[00:39:46] 
[00:39:46] ------------------------------------------
[00:39:46] stderr:
[00:39:46] ------------------------------------------
[00:39:46] 
[00:39:46] ------------------------------------------
[00:39:46] 
[00:39:46] thread '[ui] ui/feature-gate-extern_absolute_paths.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2774:8
[00:39:46] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:39:46] 
[00:39:46] 
[00:39:46] failures:
[00:39:46]     [ui] ui/feature-gate-extern_absolute_paths.rs
```

</details>

File: src/libsyntax/feature_gate.rs [hunk 4]
```diff
@@ -430,6 +430,9 @@ declare_features! (
 
     // generic associated types (RFC 1598)
     (active, generic_associated_types, "1.23.0", Some(44265)),
+
+    // Resolve absolute paths as paths from other crates
+    (active, extern_absolute_paths, "1.23.0", Some(44660)),
```
Comment (kennytm) [inline src/libsyntax/feature_gate.rs]: This should be 1.24.0?

File: src/test/ui/feature-gate-extern_absolute_paths.rs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {}
```
Comment (kennytm) [inline src/test/ui/feature-gate-extern_absolute_paths.rs]: Something is missing?

File: src/librustc_resolve/lib.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: There's no such existing function.
(Also there are too many "crate roots" with different properties now, it's better to be explicit.)

**Pull request merged**